docker/container-structure-test.yaml

```diff
@@ -0,0 +1,37 @@
+# See the file "COPYING" in the main distribution directory for copyright.
+
+# This file contains a test configuration for
+# https://github.com/GoogleContainerTools/container-structure-test.
+
+schemaVersion: 2.0.0
+
+commandTests:
+  - name: zeek can be run
```

Comment:
Is `container-structure-test` a known entity? I'm wondering whether we need the extra dependency here, since this looks pretty doable via `btest`. These could basically all be `@TEST-EXEC docker run -it zeekurity/zeek:something <command>` with the usual output validation, no? Perhaps with a `@TEST-REQUIRES` to check docker availability or some such.

Reply:
I have used `container-structure-test` before, and it seems to come from a [reputable source](https://github.com/GoogleContainerTools/container-structure-test), but I was wondering about using BTest as well. The reason I decided against that we'd some way to install BTest in an outer environment, and then start docker containers; in CI the outer environment might already be a container and I am not a fan of Docker-in-Docker.

I'll leave this open for now while I look into it in more detail.

Reply:
Sound good. Fwiw, I took a quick look at `container-structure-test` and it seems to use docker under the hood. Regarding the `btest` installation: if we run this build setup out of a full Zeek source tree we always have `btest` available via `auxil/btest`, just like the `btest` invocations in Zeek's own testsuite.

Reply:
> Sound good. Fwiw, I took a quick look at `container-structure-test` and it seems to use docker under the hood. Regarding the `btest` installation: if we run this build setup out of a full Zeek source tree we always have `btest` available via `auxil/btest`, just like the `btest` invocations in Zeek's own testsuite.

Sorry for being unclear.

The complexity comes from the need to handle at least two cases:

1. BTest run on a developer machine

    - build and tag image with some test tag in the user's namespace, e.g., `localhost/zeek-test:latest`
    - btests for container can then contain e.g., `# @TEST-EXEC: docker run --rm localhost/zeek-test:latest zeek -v` and matching checks for stdout and stderr mirroring what we currently have in the container tests in this patch

2. BTest run in CI, e.g., Github action so we can use `docker/build-push-action` to both build and publish image; this allows us to test the image we would publish and avoid adding more jobs to Cirrus CI which already requires a lot of time to schedule inside the given quotas.

    - jobs runs in VM, e.g., `ubuntu-latest`
    - check out Zeek sources with `actions/checkout`
    - use e.g., `actions/setup-python` and `BSFishy/pip-action` to set up BTest environment
    - build, tag, and load image with e.g., `docker/build-push-action`
    - since we do not have a full Zeek build only run the container btests, e.g., with `btest container/`. This requires that the `docker` command is available in the environment of the build step; I am not sure of that.

What we need to do for (2) seems pretty complicated if it can work at all.

Alternatively it might be possible to instead build another image on top of the image we would publish; this image would need to contain at least the container btests and then run commands inside that image. We could then run the tests inside that image, but would need a way for the `docker` command to work from inside a container (it might be possible to provide a test wrapper around `docker` which makes it a noop for dind). It might also be possible to offload some of that work to infra in Cirrus CI and publish from there, but I personally had some issues correctly using their build service with containers performing e.g., `COPY`, see e.g., this note from [their docs](https://cirrus-ci.org/guide/docker-builder-vm/#dockerfile-as-a-ci-environment)

>⚠️ Danger of using COPY and ADD instructions
> 
> Cirrus doesn't include files added or copied into a container image in the cache key. This means that for a public repository a potential bad actor can create a PR with malicious scripts included into a container, wait for it to be cached and then reset the PR so it looks harmless.

Since these tests just perform basic checks (e.g., command exists and produces reasonably looking output) which I would not expect to be broken frequently on something already tested much more extensively in a "real testsuite" executed in CI I felt that instead of trying to solve all these issues it might be easier to just set up something which can more easily run in CI (so people usually do not even need to know about these tests).

Reply:
After typing that response I remembered another possible approach: do not write any container tests, instead build the image, and when running it somehow mount the existing test tree into it and execute that. This is probably expensive and excessive.

Reply:
Thanks for the details Benjamin! In the end it's up to you. I've dug into this a bit today and I noticed the following: `container-structure-test` relies on docker as well, at least with its default driver. If you use the Plex action, it actually ends up using Docker to download `container-structure-test` and run the thing:

https://github.com/zeek/zeek/pull/1739/checks?check_run_id=3672951554#step:6:7

Also, maybe I'm overlooking something, but it's not hard to incorporate `btest`. I gave this a shot using your branch in my own Zeek fork, with an extra commit on top. If you want to take a look:

https://github.com/ckreibich/zeek/commit/45a536db139e3aaa722eeea129348557d8327b2d
https://github.com/ckreibich/zeek/runs/3682053557?check_suite_focus=true#step:6:1

Overall I come out in favor of `btest` because I can now run the testing locally without needing an extra tool.

Reply:
Ups, that looked way simpler than I imagined if we can avoid putting it into the default BTest suite.

> Overall I come out in favor of btest because I can now run the testing locally without needing an extra tool.

Okay, let's do that then.